### ISSUE TITLE
frozen_mutation_test: test_writing_and_reading: Mutations differ

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylla#9483]
Unit test flakiness seen in https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/4055/artifact/testlog/x86_64_dev/frozen_mutation_test.195.log 
[frozen_mutation_test.195.log](https://github.com/scylladb/scylla/files/7345641/frozen_mutation_test.195.log)

Scylla version 33f8ec09df309b055766aec5c2845f0c4e525bd4 (that was promoted later)

```diff
$ test/lib/mutation_diff /tmp/frozen_mutation_test.195.log | cat
--- /dev/fd/63	2021-10-14 14:30:49.308466694 +0300
+++ /dev/fd/62	2021-10-14 14:30:49.308466694 +0300
@@ -1,4 +1,4 @@
-./test/lib/mutation_assertions.hh(108): fatal error: in "test_writing_and_reading": Mutations differ, expected 
+ ...but got: 
 {table: 'ks.cf', key: {'pk': 1806000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000, token: -9205109811216406580}, mutation_partition: {
   tombstone: {tombstone: timestamp=-9223372036854775806, deletion_time=2},
   range_tombstones: {
@@ -68,7 +68,7 @@
       'v9': { atomic_cell{shards: {global_shard id: a6afc528-3152-4a88-b0d4-eb6dc5b4bd5c value: -81 clock: 840},ts=-9223372036854775737,expiry=-1,ttl=0} },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       marker: {row_marker: -9223372036854775737 0 0},
       position: {
@@ -97,7 +97,7 @@
       'v60': { atomic_cell{DEAD,ts=-9223372036854775745,deletion_time=1} },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       marker: {row_marker: -9223372036854775736 0 0},
       position: {
@@ -169,7 +169,7 @@
       'v9': { atomic_cell{DEAD,ts=-9223372036854775746,deletion_time=0} },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       tombstone: {row_tombstone: {tombstone: timestamp=-9223372036854775775, deletion_time=0}{shadowable tombstone: none}},
       position: {
@@ -189,7 +189,7 @@
       },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       marker: {row_marker: -9223372036854775737 0 0},
       tombstone: {row_tombstone: {tombstone: timestamp=-9223372036854775776, deletion_time=2}{shadowable tombstone: none}},
@@ -273,7 +273,7 @@
       'v9': { atomic_cell{DEAD,ts=-9223372036854775747,deletion_time=1} },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       marker: {row_marker: -9223372036854775736 0 0},
       position: {
@@ -316,7 +316,7 @@
       },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       marker: {row_marker: -9223372036854775736 0 0},
       tombstone: {row_tombstone: {tombstone: timestamp=-9223372036854775777, deletion_time=1}{shadowable tombstone: none}},
@@ -327,7 +327,7 @@
       },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       tombstone: {row_tombstone: {tombstone: none}{shadowable tombstone: timestamp=-9223372036854775785, deletion_time=1}},
       position: {
@@ -358,7 +358,7 @@
       'v9': { atomic_cell{shards: {global_shard id: a31597c0-44b5-4fba-ba99-7902e3a2a0ab value: 13 clock: 12801},ts=-9223372036854775735,expiry=-1,ttl=0} },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       position: {
         bound_weight: 0,
@@ -406,7 +406,7 @@
       'v7': { atomic_cell{DEAD,ts=-9223372036854775746,deletion_time=0} },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       position: {
         bound_weight: 0,
@@ -436,7 +436,7 @@
       'v9': { atomic_cell{DEAD,ts=-9223372036854775746,deletion_time=1} },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       tombstone: {row_tombstone: {tombstone: timestamp=-9223372036854775775, deletion_time=0}{shadowable tombstone: none}},
       position: {
@@ -546,7 +546,7 @@
       },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       position: {
         bound_weight: 0,
@@ -575,7 +575,7 @@
       'v9': { atomic_cell{DEAD,ts=-9223372036854775746,deletion_time=0} },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       marker: {row_marker: dead -9223372036854775765 0},
       tombstone: {row_tombstone: {tombstone: timestamp=-9223372036854775776, deletion_time=2}{shadowable tombstone: none}},
@@ -586,7 +586,7 @@
       },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       marker: {row_marker: -9223372036854775735 1 0},
       position: {
@@ -604,7 +604,7 @@
       'v60': { atomic_cell{DEAD,ts=-9223372036854775746,deletion_time=1} },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       position: {
         bound_weight: 0,
@@ -627,7 +627,6 @@
       cont: true,
       dummy: false,
       marker: {row_marker: -9223372036854775737 1 0},
-      tombstone: {row_tombstone: {tombstone: none}{shadowable tombstone: timestamp=-9223372036854775786, deletion_time=0}},
       position: {
         bound_weight: 0,
         'ck1': eb02000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000,
@@ -676,7 +675,7 @@
       'v9': { atomic_cell{shards: {global_shard id: 0723ed29-ffbb-4f47-82d8-aa1867fc930d value: -12 clock: 859}, {global_shard id: 1802b090-e896-460d-bbd3-e5acd5c6cc8e value: 69 clock: 13989}, {global_shard id: 29109aea-9f9d-4b0b-a7c9-3d331f8294d9 value: -3 clock: 1395}, {global_shard id: 82f3f4c4-68d9-4136-a853-feb80bc99ac6 value: 84 clock: 14134}, {global_shard id: a6afc528-3152-4a88-b0d4-eb6dc5b4bd5c value: -34 clock: 8930}, {global_shard id: d66a70d9-9e0f-4368-b6a3-23453413ac13 value: -6 clock: 9409},ts=-9223372036854775736,expiry=-1,ttl=0} },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       marker: {row_marker: -9223372036854775736 0 0},
       position: {
@@ -730,7 +729,7 @@
       },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       marker: {row_marker: dead -9223372036854775765 0},
       tombstone: {row_tombstone: {tombstone: none}{shadowable tombstone: timestamp=-9223372036854775787, deletion_time=1}},
@@ -741,7 +740,7 @@
       },
     },
     {
-      cont: false,
+      cont: true,
       dummy: false,
       marker: {row_marker: -9223372036854775735 1 1},
       tombstone: {row_tombstone: {tombstone: timestamp=-9223372036854775775, deletion_time=2}{shadowable tombstone: none}},
@@ -754,3 +753,4 @@
   ]
 }
 }
+test/boost/frozen_mutation_test.cc(0): Leaving test case "test_writing_and_reading"; testing time: 282509us
```

**Comment by denesb** [comment]
Upon looking again I think the problem was entirely in the test and the original fix (6a76e127683355e4d621dfd4905231884b36f7be) should have fixed this. I think the problem was in this code: https://github.com/scylladb/scylla/blob/9d4262e264efd6d92de9c40545f59c85c840de0d/test/lib/mutation_source_test.cc#L2310-L2339
Here it is possible that the same ckey will be generated twice. On the first pass a shadowable tombstone and possibly a row marker is applied to the row. On the second pass only a row marker is applied but no shadowable tombstone. Before the fix applying the row marker didn't trigger a shadow check so it is possible that the row had both a row marker and a shadowable tombstone shadowed by it triggering the difference seen above. This has a really small chance, but it can happen.

I also looked at the row tombstone code and I don't see any other sequence that could lead to this.

I will close, we can reopen if we ever see this again (now with the fix in).

**Comment by avikivity** [comment]
Backported to 5.0 by @tgrabiec, now further backported to 4.6.
